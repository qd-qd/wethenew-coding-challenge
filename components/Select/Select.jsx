import PropTypes from 'prop-types';
import styles from './Select.module.css';

const Select = ({ label, values, valueSelected, onChange }) => {
  return (
    <label
      htmlFor={`${label}-select`}
      className={styles.label}
      data-testid="label"
    >
      {label}
      <div className={styles.select__container}>
        <select
          id={`${label}-select`}
          className={styles.select}
          onChange={e => onChange(e.target.value)}
          value={valueSelected}
          data-testid="select"
        >
          {values.map(({ label: optionLabel, value }) => (
            <option value={value} key={value} data-testid={`option-${value}`}>
              {optionLabel}
            </option>
          ))}
        </select>
      </div>
    </label>
  );
};

Select.defaultProps = {
  onChange: () => {},
};

Select.propTypes = {
  label: PropTypes.string.isRequired,
  values: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
        .isRequired,
    }),
  ).isRequired,
  valueSelected: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    .isRequired,
  onChange: PropTypes.func,
};

export default Select;
