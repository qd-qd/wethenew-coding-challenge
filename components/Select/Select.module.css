.label {
  text-transform: uppercase;
  color: var(--theme-color-text-secondary);
  font-weight: var(--theme-font-weight-regular);
  font-size: 1.1rem;
  width: 100%;
}

.select__container {
  margin-top: 0.5rem;
}

.select__container > select {
  display: inline-block;
  min-width: 150px;
  cursor: pointer;
  padding: 10px 15px;
  outline: 0;
  border: 1px solid var(--theme-color-grey-400);
  border-radius: var(--theme-shape-radius);
  background: var(--theme-color-background);
  color: var(--theme-color-text-primary);
  appearance: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  font-size: 1.1rem;
}

.select__container > select::-ms-expand {
  display: none;
}

.select__container > select:hover,
.select__container > select:focus {
  background: var(--theme-color-grey-200);
}

.select__container > select:disabled {
  opacity: 0.5;
  pointer-events: none;
}

.select {
  position: relative;
  display: inline-block;
  margin-bottom: 15px;
  width: 100%;
}

.select:after {
  position: absolute;
  content: '';
  top: 14px;
  right: 10px;
  width: 0;
  height: 0;
  border: 6px solid transparent;
  border-color: #fff transparent transparent transparent;
}

@media (min-width: 35em) {
  .label {
    width: unset;
  }

  .select {
    min-width: 150px;
  }
}
