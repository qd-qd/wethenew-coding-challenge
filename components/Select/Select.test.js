import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Select from './Select';

let props = {};

beforeEach(() => {
  props = {
    label: 'fake label',
    values: [
      {
        label: 'label-1',
        value: '1',
      },
      {
        label: 'label-2',
        value: '2',
      },
      {
        label: 'label-3',
        value: '3',
      },
      {
        label: 'label-4',
        value: '4',
      },
    ],
    valueSelected: '3',
    onChange: jest.fn(() => {}),
  };
});

describe('Select', () => {
  test('displays greeting', () => {
    render(
      <Select
        label={props.label}
        values={props.values}
        valueSelected={props.valueSelected}
        onChange={props.onChange}
      />,
    );

    expect(screen.getByTestId('label')).toBeDefined();
    expect(screen.getByTestId('label')).toHaveAttribute('for');
    expect(screen.getByTestId('label').htmlFor).toBe(`${props.label}-select`);

    expect(screen.getByText(props.label)).toBeDefined();

    expect(screen.getByTestId('select')).toBeDefined();
    expect(screen.getByTestId('select')).toHaveAttribute('class');
    expect(screen.getByTestId('select')).toHaveAttribute('id');
    expect(screen.getByTestId('select').id).toBe(`${props.label}-select`);

    props.values.forEach(({ value, label }) => {
      expect(screen.getByTestId(`option-${value}`)).toBeDefined();
      expect(screen.getByTestId(`option-${value}`).value).toBe(value);
      expect(screen.getByText(label)).toBeDefined();
    });

    // @dev: select element displays the correct value
    expect(screen.getByDisplayValue(props.values[2].label)).toBeDefined();
  });

  test('is interactive', async () => {
    render(
      <Select
        label={props.label}
        values={props.values}
        valueSelected={props.valueSelected}
        onChange={props.onChange}
      />,
    );

    expect(screen.getByDisplayValue(props.values[2].label)).toBeDefined();
    fireEvent.change(
      screen.getByTestId('select', {
        target: { value: '4' },
      }),
    );
    expect(props.onChange).toHaveBeenCalledTimes(1);
    fireEvent.change(
      screen.getByTestId('select', {
        target: { value: '1' },
      }),
    );
    expect(props.onChange).toHaveBeenCalledTimes(2);
  });

  test("doesn't break snapshot", () => {
    const { container } = render(
      <Select
        label={props.label}
        values={props.values}
        valueSelected={props.valueSelected}
        onChange={props.onChange}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
