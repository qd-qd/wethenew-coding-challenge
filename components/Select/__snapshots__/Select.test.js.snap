// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select doesn't break snapshot 1`] = `
<div>
  <label
    class="label"
    data-testid="label"
    for="fake label-select"
  >
    fake label
    <div
      class="select__container"
    >
      <select
        class="select"
        data-testid="select"
        id="fake label-select"
      >
        <option
          data-testid="option-1"
          value="1"
        >
          label-1
        </option>
        <option
          data-testid="option-2"
          value="2"
        >
          label-2
        </option>
        <option
          data-testid="option-3"
          value="3"
        >
          label-3
        </option>
        <option
          data-testid="option-4"
          value="4"
        >
          label-4
        </option>
      </select>
    </div>
  </label>
</div>
`;
